common/functions/src/scalars/dates/week_date.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::marker::PhantomData;
+use std::ops::Sub;
+
+use common_datavalues::chrono::DateTime;
+use common_datavalues::chrono::Datelike;
+use common_datavalues::chrono::Duration;
+use common_datavalues::chrono::TimeZone;
+use common_datavalues::chrono::Utc;
+use common_datavalues::prelude::*;
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::scalars::Function;
+
+#[derive(Clone, Debug)]
+pub struct WeekFunction<T, R> {
+    display_name: String,
+    t: PhantomData<T>,
+    r: PhantomData<R>,
+}
+
+pub trait WeekResultFunction<R> {
+    fn return_type() -> Result<DataType>;
+    fn to_number(_value: DateTime<Utc>, mode: Option<u64>) -> R;
+    fn to_constant_value(_value: DateTime<Utc>, mode: Option<u64>) -> DataValue;
+}
+
+#[derive(Clone)]
+pub struct ToStartOfWeek;
+
+impl WeekResultFunction<u32> for ToStartOfWeek {
+    fn return_type() -> Result<DataType> {
+        Ok(DataType::Date16)
+    }
+    fn to_number(value: DateTime<Utc>, mode: Option<u64>) -> u32 {
+        let week_mode = mode.unwrap_or(0);
+        let mut weekday = value.weekday().number_from_monday();
+        if (week_mode & 1) == 0 {
+            weekday = value.weekday().number_from_sunday();
```

Comment:
If week_mode is 1,should  use monday first?
```
   bool monday_first_mode = week_mode & static_cast<UInt8>(WeekModeFlag::MONDAY_FIRST);
        if (monday_first_mode)
        {
            return toFirstDayNumOfWeek(v);
        }
```

Reply:
my mistake

Reply:
I find clickhouse doc, if week_mode is 1,First day of week is Monday. https://clickhouse.com/docs/zh/sql-reference/functions/date-time-functions/#toweekdatemode

Reply:
shouldn't I have 1 in the code? This is magic number?

Reply:
No, just a little weird.

Default weekday should be  `number_from_sunday` because of the week_mode defaults to 0.

Reply:
I understand.I will modify it right away